draft-rescorla-tls-esni.md

```diff
@@ -312,25 +317,27 @@ encrypted_sni
 
 In order to send an encrypted SNI, the client MUST first select one of
 the server ESNIKeyShareEntry values and generate an (EC)DHE share in the
-matching group. This share is then used for the client's "key_share"
-extension and will be used to derive both the SNI encryption
-key and the (EC)DHE shared secret which is used in the TLS key schedule.
-This has two important implications:
+matching group. This share will then be sent to the server in the EncryptedSNI
+extension and used to derive the SNI encryption key. It does not affect the 
+(EC)DHE shared secret used in the TLS key schedule.
 
-- The client MUST only provide one KeyShareEntry
+This has one important implication: The client-facing server is committing to 
+support every group in the ESNIKeys list (see below for server behavior). 
+The (backend) server terminating the TLS connection makes no such commitment,
+as its key exchange algorithm selection is unaffected.
 
-- The server is committing to support every group in the
-  ESNIKeys list (see below for server behavior).
-
-The SNI encryption key is computed from the DH shared secret Z as
-follows:
+Let Z be the DH shared secret derived from a key share in ESNIKeys and the 
+corresponding client share in EncryptedSNI.entry. The SNI encryption key is 
+Z as follows:
 
 ~~~~
    Zx = HKDF-Extract(0, Z)
-   key = HKDF-Expand-Label(Zx, "esni key", Hash(ClientHello.Random), key_length)
-   iv = HKDF-Expand-Label(Zx, "esni iv", Hash(ClientHello.Random), iv_length)
+   key = HKDF-Expand-Label(Zx, "esni key", ESNIContents, key_length)
+   iv = HKDF-Expand-Label(Zx, "esni iv", ESNIContents, iv_length)
 ~~~~
 
+where ESNIContents = Hash(record_digest || ClientHello.KeyShareClientHello || ClientHello.Random),
+and Hash is the hash function associated with the HKDF instantiation.
```

Comment:
I think this is not quite what we want.

I think rather the hash should be:

```Hash(record_digest || EncryptedSNI.entry || ClientHello.Random)```

And then you want to have the AAD include ClientHello.key_shares.

Reply:
EncryptedSNI.entry is already mixed in by virtue of being used to derive the KDF input secret. Putting ClientHello.KeyShareClientHello here or in the AAD has the same affect on malleability.